custom/auth/ante/fee.go

```diff
@@ -220,10 +257,22 @@ func (fd FeeDecorator) checkTxFee(ctx sdk.Context, tx sdk.Tx, taxes sdk.Coins) (
 		}
 
 		requiredFees := requiredGasFees.Add(taxes...)
+		allFees := requiredFees.Add(nonTaxableTaxes...)
 
 		// Check required fees
 		if !requiredFees.IsZero() && !feeCoins.IsAnyGTE(requiredFees) {
-			return 0, errorsmod.Wrapf(sdkerrors.ErrInsufficientFee, "insufficient fees; got: %q, required: %q = %q(gas) + %q(stability)", feeCoins, requiredFees, requiredGasFees, taxes)
+			// we don't have enough for tax and gas fees. But do we have enough for gas alone?
+			if !requiredGasFees.IsZero() && !feeCoins.IsAnyGTE(requiredGasFees) {
+				return 0, false, false, errorsmod.Wrapf(sdkerrors.ErrInsufficientFee, "insufficient fees; got: %q, required: %q = %q(gas) + %q(stability)", feeCoins, requiredFees, requiredGasFees, taxes)
+			}
+
+			// we have enough for gas fees but not for tax fees
+			reverseCharge = true
+		}
+
+		if !allFees.IsZero() && feeCoins.IsAnyGTE(allFees) {
+			// we have enough for all fees
+			refundNonTaxableTaxes = true
 		}
 	}
```

Comment:
for my understanding:
- we apply `reverse_charge` when txFee is just enough to pay for gas fee: `gasFee < txFee < gasFee + taxes` 
- we add `refundNonTaxableTaxes` when txFee cover all the fees `txFee > gasFee + taxes + nontaxableTaxes`

Their is a case, where `txFee > gasFee + taxes`, but `txFee < allFee`. That makes `refundNonTaxableTaxes` false.

I imagine their is a case, where there are two msgs, `MsgBank` and `ExecuteContract`, which should be non-taxable. Then, if the user set `gasFee + Taxes < txFee < allFee`, the`nonTaxableTaxes` will not be deducted.

The logic here is good already (we just ignore that case, considering it the extra fee).  

Comment here is my wonder if we should return some portion of this fee since it the tx contain `nonTaxableTaxes`?

Reply:
`The only option I see that could be implemented would be to check for the sent gasLimit (gasWanted) and refund everything that is above (gasWanted * gasPrice) + tax. Would that make sense to do or rather not?` 

i think your implementation is good already, lets stick with that. The CLI is provided with the estimate function already, so if user happens to put large fee, we should keep the current handling since that is expected to charge all the fee they send.